README
NAME
    Lingua::JA::Romanize::Japanese - Romanization of Japanese language

SYNOPSIS
        use Lingua::JA::Romanize::Japanese;

        my $conv = Lingua::JA::Romanize::Japanese->new();
        my $roman = $conv->char( $kanji );
        printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $kanji, $roman );

        my @array = $conv->string( $string );
        foreach my $pair ( @array ) {
            my( $raw, $ruby ) = @$pair;
            if ( defined $ruby ) {
                printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
            } else {
                print $raw;
            }
        }

DESCRIPTION
    Japanese language is written with a mix of Kanji and Kana characters.
    Most of Kanji characters used in Japan were imported from China. Two
    types of Kana characters, called Katakana and Hiragana, were created in
    Japan. Kana characters are general terms for the syllabic Japanese
    scripts.

METHODS
  $conv = Lingua::JA::Romanize::Japanese->new();
    This constructer methods returns a new object with its dictionary
    cached.

  $roman = $conv->char( $kanji );
    This method returns romanized letters of a Japanese character. It
    returns undef when $Kana is not a valid Japanese character. The
    argument's encoding must be UTF-8. Both of Kanji and Kana characters are
    allowed.

  $roman = $conv->chars( $string );
    This method returns romanized letters of Japanese characters.

  @array = $conv->string( $string );
    This method returns a array of referenced arrays which are pairs of
    Japanese chacater(s) and its romanized letters.

        $array[0]           # first Japanese character(s)'s pair (array)
        $array[1][0]        # secound Japanese character(s) itself
        $array[1][1]        # its romanized letters

DICTIONARY
    This module's Japanese to roman mapping table is based on the dictionary
    of SKK which is a Japanese input method on Emacs. It was designed by Dr.
    Masahiko Sato and created in 1987. SKK is an abbreviation of 'Simple
    Kana to Kanji conversion program'.

MODULE DEPENDENCIES
    DB_File module is required.

SEE ALSO
    Lingua::JA::Romanize::Kana Lingua::JA::Romanize::Juman
    Lingua::JA::Romanize::MeCab Lingua::ZH::Romanize::Pinyin
    Lingua::KO::Romanize::Hangul

    http://www.kawa.net/works/perl/romanize/romanize-e.html

    http://openlab.jp/skk/

AUTHOR
    Yusuke Kawasaki http://www.kawa.net/

COPYRIGHT
    Copyright (c) 2006 Yusuke Kawasaki. All rights reserved.

LICENSE
    This dictionary is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by the
    Free Software Foundation; either versions 2, or (at your option) any
    later version.



Changes
# Lingua::JA::Romanize::* modules Changes

2006/11/06 (0.14)
    - test scripts updated for tests passed.
    - MeCab 0.93 with ipadic-2.7.0-20060707 tested.

2006/05/08 (0.13)
    - new sub classes: MeCab::UTF8 in Lingua::JA::Romanize::MeCab
    - t/03_juman.t tests $PATH to find juman command.
    - t/04_mecab.t supports dictionary charset detection.
    - perltidy-ed.

2006/04/04 (0.12)
    - new module: Lingua::JA::Romanize::Juman
    - new module: Lingua::JA::Romanize::MeCab

2006/03/26 (0.11)
    - first release

# http://www.kawa.net/works/perl/romanize/romanize-e.html       (English)
# http://www.kawa.net/works/perl/romanize/romanize.html#history (Japanese)


lib/Lingua/JA/Romanize/Kana.pm
=head1 NAME

Lingua::JA::Romanize::Kana - Romanization of Japanese Hiragana/Katakana

=head1 SYNOPSIS

    use Lingua::JA::Romanize::Kana;

    my $conv = Lingua::JA::Romanize::Kana->new();
    my $roman = $conv->char( $kana );
    printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $kana, $roman );

    my @array = $conv->string( $string );
    foreach my $pair ( @array ) {
        my( $raw, $ruby ) = @$pair;
        if ( defined $ruby ) {
            printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
        } else {
            print $raw;
        }
    }

=head1 DESCRIPTION

Hiragana and Katakana are general terms for the syllabic Japanese scripts.

=head1 METHODS

=head2 $conv = Lingua::JA::Romanize::Kana->new();

This constructer methods returns a new object with its dictionary cached.

=head2 $roman = $conv->char( $Kana );

This method returns romanized letters of a Kana character.
It returns undef when $Kana is not a valid Kana character.
The argument's encoding must be UTF-8.
Both of Hiragana or Katakana characters are allowed.
But Kanji character is not supported by this module.
See L<Lingua::JA::Romanize::Japanese>.

=head2 $roman = $conv->chars( $string );

This method returns romanized letters of Kana characters.

=head2 @array = $conv->string( $string );

This method returns a array of referenced arrays
which are pairs of a Kana chacater and its romanized letters.

    $array[0]           # first Kana character's pair (array)
    $array[1][0]        # secound Kana character itself
    $array[1][1]        # its romanized letters

=head1 SEE ALSO

L<Lingua::JA::Romanize::Japanese>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2006 Yusuke Kawasaki. All rights reserved.

This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=cut
# ----------------------------------------------------------------
    package Lingua::JA::Romanize::Kana;
    use strict;
    use vars qw( $VERSION );
    $VERSION = "0.12";
# ----------------------------------------------------------------
    my $KANA_MAP = [qw(
            xa  a   xi  i   xu  u   xe  e   xo  o   ka  ga  ki  gi  ku
        gu  ke  ge  ko  go  sa  za  shi ji  su  zu  se  ze  so  zo  ta
        da  chi ji  xtu tsu zu  te  de  to  do  na  ni  nu  ne  no  ha
        ba  pa  hi  bi  pi  fu  bu  pu  he  be  pe  ho  bo  po  ma  mi
        mu  me  mo  xya ya  xyu yu  xyo yo  ra  ri  ru  re  ro  xwa wa
        wi  we  wo  n   vu  ka  ke
    )];
# ----------------------------------------------------------------
sub new {
    my $package = shift;
    my $self = {@_};
    bless $self, $package;
    $self;
}
# ----------------------------------------------------------------
sub char {
    my $self = shift;
    my $char = shift;
    my( $c1, $c2, $c3, $c4 ) = unpack("C*",$char);
    return if ( ! defined $c3 || defined $c4 );
    my $ucs2 = (($c1 & 0x0F)<<12) | (($c2 & 0x3F)<<6) | ($c3 & 0x3F);
    return if ( $ucs2 < 0x3041 );
    return if ( $ucs2 > 0x3093 && $ucs2 < 0x30A1 );
    return if ( $ucs2 > 0x30F6 );
    my $offset = ( $ucs2 - 0x3041 ) % 96;
    $KANA_MAP->[$offset];
}
# ----------------------------------------------------------------
sub chars {
    my $self = shift;
    my @array = $self->string( shift );
    join( "", map {$#$_>0 ? $_->[1] : $_->[0]} @array );
}
# ----------------------------------------------------------------
sub string {
    my $self = shift;
    my $src = shift;
    my $array = [];

    while ( $src =~ /(\xE3[\x80-\xBF]{2})|([^\xE3]+)/sg ) {
        if ( defined $1 ) {
            my $pair = [ $1 ];
            my $roman = $self->char( $1 );
            $pair->[1] = $roman if defined $roman;
            push( @$array, $pair );
        } else {
            push( @$array, [ $2 ] );
        }
    }

    $self->normalize( $array );
}
# ----------------------------------------------------------------
sub normalize {
    my $self = shift;
    my $array = shift;

    for( my $i=0; $i<$#$array; $i ++ ) {
        next if ( scalar @{$array->[$i]} < 2 );
        next if ( scalar @{$array->[$i+1]} < 2 );
        my $this = $array->[$i]->[1];
        my $next = $array->[$i+1]->[1];
        if ( $this eq "n" && $next =~ /^[bmp]/ ) {
            $array->[$i]->[1] = "m";
        } elsif ( $this eq "xtu" && $next =~ /^([^aiueo])/ ) {
            my $head = $1;
            $head = "t" if ( $head eq "c" );
            $array->[$i+1]->[0] = $array->[$i]->[0].$array->[$i+1]->[0];
            $array->[$i+1]->[1] = $head.$next;
            $array->[$i] = undef;
        } elsif ( $this =~ /i$/ && $next =~ /^xy/ ) {
            my $head = ( $this =~ /^(.*)i$/ )[0];
            my $tail = ( $next =~ /^x(y.*)$/ )[0];
            $tail =~ s/^y// if ( $head =~ /(.h|^j)$/ );
            $array->[$i+1]->[0] = $array->[$i]->[0].$array->[$i+1]->[0];
            $array->[$i+1]->[1] = $head.$tail;
            $array->[$i] = undef;
        } elsif ( $this =~ /o$/ && $next =~ /^x?o$/ ) {
            $array->[$i+1]->[0] = $array->[$i]->[0].$array->[$i+1]->[0];
            $array->[$i+1]->[1] = $this."h";
            $array->[$i] = undef;
        } elsif ( $this eq "vu" && $next =~ /^x([aiueo])$/ ) {
            my $tail = $1;
            $array->[$i+1]->[0] = $array->[$i]->[0].$array->[$i+1]->[0];
            $array->[$i+1]->[1] = "v".$tail;
            $array->[$i] = undef;
        }
    }
    $array = [ grep { ref $_ } @$array ];

    for( my $i=0; $i<=$#$array; $i ++ ) {
        next if ( scalar @{$array->[$i]} < 2 );
        $array->[$i]->[1] =~ s/^x//;
    }

    @$array;
}
# ----------------------------------------------------------------
;1;
# ----------------------------------------------------------------


lib/Lingua/JA/Romanize/Juman.pm
=head1 NAME

Lingua::JA::Romanize::Juman - Romanization of Japanese language with JUMAN

=head1 SYNOPSIS

    use Lingua::JA::Romanize::Juman;

    my $conv = Lingua::JA::Romanize::Juman->new();
    my $roman = $conv->char( $kanji );
    printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $kanji, $roman );

    my @array = $conv->string( $string );
    foreach my $pair ( @array ) {
        my( $raw, $ruby ) = @$pair;
        if ( defined $ruby ) {
            printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
        } else {
            print $raw;
        }
    }

=head1 DESCRIPTION

This is JUMAN version of L<Lingua::JA::Romanize::Japanese> module.
Both of JUMAN and its Perl binding, Juman.pm, are required.

=head1 SEE ALSO

L<Lingua::JA::Romanize::Japanese>

http://www.kc.t.u-tokyo.ac.jp/nl-resource/juman.html (Japanese)

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2006 Yusuke Kawasaki. All rights reserved.

This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=cut
# ----------------------------------------------------------------
    package Lingua::JA::Romanize::Juman;
    use strict;
    use Carp;
    use Juman;
    use Encode;
    use Lingua::JA::Romanize::Kana;
    use vars qw( $VERSION );
    $VERSION = "0.12";
# ----------------------------------------------------------------
    if ( $] > 5.008 ) {
        require Encode;
    } else {
        local $@;
        eval { require Jcode; };
        die "Jcode.pm is required on Perl $]\n" if $@;
    }
# ----------------------------------------------------------------
sub new {
    my $package = shift;
    my $self = {};
    $self->{juman} = Juman->new( @_ );
    $self->{kana} = Lingua::JA::Romanize::Kana->new();
    $self->{jcode} = Jcode->new("") unless ( $] > 5.008 );
    bless $self, $package;
    $self;
}
# ----------------------------------------------------------------
sub char {
    my $self = shift;
    my $src = shift;
    my $roman = $self->{kana}->char($src);
    return $roman if $roman;
    $src = $self->utf8_to_eucjp( $src );
    my $result = $self->{juman}->analysis( $src ) or return;
    my $node = $result->mrph(0) or return;
    my $kana = $node->yomi() or return;
    $kana = $self->eucjp_to_utf8( $kana );
    my @array = grep {$#$_>0} $self->{kana}->string($kana);
    return unless scalar @array;
    join( "", map {$_->[1]} @array );
}
# ----------------------------------------------------------------
sub chars {
    my $self = shift;
    my @array = $self->string( shift );
    join( " ", map {$#$_>0 ? $_->[1] : $_->[0]} @array );
}
# ----------------------------------------------------------------
sub string {
    my $self = shift;
    my $src = shift;
    Encode::from_to( $src, "UTF-8", "EUC-JP" );
    my $result = $self->{juman}->analysis( $src );
    my $array = [];

    foreach my $node ( $result->mrph() ) {
        my $midasi = $node->midasi();
        $midasi =~ s/^\\//;
        my $hinsi = $node->hinsi_id();
        my $kana = $node->yomi() if ( $hinsi != 1 );
        $midasi = $self->eucjp_to_utf8( $midasi ) if defined $midasi;
        $kana = $self->eucjp_to_utf8( $kana ) if defined $kana;
        my @array = $self->{kana}->string($kana) if $kana;
        my $roman = join( "", map {$_->[1]} grep {$#$_>0} @array ) if scalar @array;
        my $pair = $roman ? [ $midasi, $roman ] : [ $midasi ];
        push( @$array, $pair );
    }

    $self->{kana}->normalize( $array );
}
# ----------------------------------------------------------------
sub utf8_to_eucjp {
    my $self = shift;
    my $src = shift;
    if ( $] > 5.008 ) {
        Encode::from_to( $src, "UTF-8", "EUC-JP" );
    } else {
        $src = $self->{jcode}->set( \$src, "utf8" )->euc();
    }
    $src;
}
# ----------------------------------------------------------------
sub eucjp_to_utf8 {
    my $self = shift;
    my $src = shift;
    if ( $] > 5.008 ) {
        Encode::from_to( $src, "EUC-JP", "UTF-8" );
    } else {
        $src = $self->{jcode}->set( \$src, "euc" )->utf8();
    }
    $src;
}
# ----------------------------------------------------------------
;1;
# ----------------------------------------------------------------


lib/Lingua/JA/Romanize/MeCab.pm
=head1 NAME

Lingua::JA::Romanize::MeCab - Romanization of Japanese language with MeCab

=head1 SYNOPSIS

    use Lingua::JA::Romanize::MeCab;

    my $conv = Lingua::JA::Romanize::MeCab->new();
    my $roman = $conv->char( $kanji );
    printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $kanji, $roman );

    my @array = $conv->string( $string );
    foreach my $pair ( @array ) {
        my( $raw, $ruby ) = @$pair;
        if ( defined $ruby ) {
            printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
        } else {
            print $raw;
        }
    }

=head1 DESCRIPTION

This is MeCab version of L<Lingua::JA::Romanize::Japanese> module.
MeCab's Perl binding, MeCab.pm, is required.

=head1 UTF-8 DICTIONARY

If MeCab's dictionary is generated with UTF8, --with-charset=utf8,
use Lingua::JA::Romanize::MeCab::UTF8->new()
instead of Lingua::JA::Romanize::MeCab->new().

=head1 SEE ALSO

L<Lingua::JA::Romanize::Japanese>

http://mecab.sourceforge.jp/ (Japanese)

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2006 Yusuke Kawasaki. All rights reserved.

This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=cut
# ----------------------------------------------------------------
package Lingua::JA::Romanize::MeCab;
use strict;
use Carp;
use MeCab;
use Lingua::JA::Romanize::Kana;
use vars qw( $VERSION );
$VERSION = "0.13";

# ----------------------------------------------------------------
sub new {
    my $package = shift;
    my $self    = {};
    &require_encode_or_jcode();
    $self->{mecab} = MeCab::Tagger->new(@_);
    $self->{kana}  = Lingua::JA::Romanize::Kana->new();
    $self->{jcode} = Jcode->new("") unless ( $] > 5.008 );
    bless $self, $package;
    $self;
}

sub char {
    my $self  = shift;
    my $src   = shift;
    my $roman = $self->{kana}->char($src);
    return $roman if $roman;
    my $pair =
      ( $self->string($src) )[0];    # need loop for nodes which have surface
    return if ( scalar @$pair == 1 );
    return $pair->[1];
}

sub chars {
    my $self  = shift;
    my @array = $self->string(shift);
    join( " ", map { $#$_ > 0 ? $_->[1] : $_->[0] } @array );
}

sub string {
    my $self  = shift;
    my $src   = $self->from_utf8(shift);
    my $array = [];

    my $node = $self->{mecab}->parseToNode($src);
    for ( ; $node ; $node = $node->{next} ) {
        next unless defined $node->{surface};
        my $midasi = $self->to_utf8( $node->{surface} );
        my $kana = ( split( /,/, $node->{feature} ) )[7];
        $kana = $self->to_utf8($kana) if defined $kana;
        my @array = $self->{kana}->string($kana) if $kana;
        my $roman = join( "", map { $_->[1] } grep { $#$_ > 0 } @array )
          if scalar @array;
        my $pair = $roman ? [ $midasi, $roman ] : [$midasi];
        push( @$array, $pair );
    }

    $self->{kana}->normalize($array);
}

sub require_encode_or_jcode {
    if ( $] > 5.008 ) {
        return if defined $Encode::VERSION;
        require Encode;
    }
    else {
        return if defined $Jcode::VERSION;
        local $@;
        eval { require Jcode; };
        Carp::croak "Jcode.pm is required on Perl $]\n" if $@;
    }
}

*from_utf8 = \&Lingua::JA::Romanize::MeCab::EUC::from_utf8;
*to_utf8 = \&Lingua::JA::Romanize::MeCab::EUC::to_utf8;

# ----------------------------------------------------------------
package Lingua::JA::Romanize::MeCab::UTF8;
use strict;
use vars qw( @ISA );
@ISA = qw( Lingua::JA::Romanize::MeCab );

sub from_utf8 {
    $_[1];              # no need to encode
}

sub to_utf8 {
    $_[1];              # no need to decode
}

# ----------------------------------------------------------------
package Lingua::JA::Romanize::MeCab::EUC;
use strict;
use vars qw( @ISA );
@ISA = qw( Lingua::JA::Romanize::MeCab );

sub from_utf8 {
    my $self = shift;
    my $src  = shift;
    if ( $] > 5.008 ) {
        Encode::from_to( $src, "UTF-8", "EUC-JP" );
    }
    else {
        $src = $self->{jcode}->set( \$src, "utf8" )->euc();
    }
    $src;
}

sub to_utf8 {
    my $self = shift;
    my $src  = shift;
    if ( $] > 5.008 ) {
        Encode::from_to( $src, "EUC-JP", "UTF-8" );
    }
    else {
        $src = $self->{jcode}->set( \$src, "euc" )->utf8();
    }
    $src;
}

# ----------------------------------------------------------------
package Lingua::JA::Romanize::MeCab::SJIS;
use strict;
use vars qw( @ISA );
@ISA = qw( Lingua::JA::Romanize::MeCab );

sub from_utf8 {
    my $self = shift;
    my $src  = shift;
    if ( $] > 5.008 ) {
        Encode::from_to( $src, "UTF-8", "CP932" );
    }
    else {
        $src = $self->{jcode}->set( \$src, "utf8" )->sjis();
    }
    $src;
}

sub to_utf8 {
    my $self = shift;
    my $src  = shift;
    if ( $] > 5.008 ) {
        Encode::from_to( $src, "CP932", "UTF-8" );
    }
    else {
        $src = $self->{jcode}->set( \$src, "sjis" )->utf8();
    }
    $src;
}

# ----------------------------------------------------------------
1;


lib/Lingua/JA/Romanize/Japanese.pm
=head1 NAME

Lingua::JA::Romanize::Japanese - Romanization of Japanese language

=head1 SYNOPSIS

    use Lingua::JA::Romanize::Japanese;

    my $conv = Lingua::JA::Romanize::Japanese->new();
    my $roman = $conv->char( $kanji );
    printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $kanji, $roman );

    my @array = $conv->string( $string );
    foreach my $pair ( @array ) {
        my( $raw, $ruby ) = @$pair;
        if ( defined $ruby ) {
            printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
        } else {
            print $raw;
        }
    }

=head1 DESCRIPTION

Japanese language is written with a mix of Kanji and Kana characters.
Most of Kanji characters used in Japan were imported from China.
Two types of Kana characters, called Katakana and Hiragana, were created in Japan.
Kana characters are general terms for the syllabic Japanese scripts.

=head1 METHODS

=head2 $conv = Lingua::JA::Romanize::Japanese->new();

This constructer methods returns a new object with its dictionary cached.

=head2 $roman = $conv->char( $kanji );

This method returns romanized letters of a Japanese character.
It returns undef when $Kana is not a valid Japanese character.
The argument's encoding must be UTF-8.
Both of Kanji and Kana characters are allowed.

=head2 $roman = $conv->chars( $string );

This method returns romanized letters of Japanese characters.

=head2 @array = $conv->string( $string );

This method returns a array of referenced arrays
which are pairs of Japanese chacater(s) and its romanized letters.

    $array[0]           # first Japanese character(s)'s pair (array)
    $array[1][0]        # secound Japanese character(s) itself
    $array[1][1]        # its romanized letters

=head1 DICTIONARY

This module's Japanese to roman mapping table is based on
the dictionary of SKK which is a Japanese input method on Emacs.
It was designed by Dr. Masahiko Sato and created in 1987.
SKK is an abbreviation of 'Simple Kana to Kanji conversion program'.

=head1 MODULE DEPENDENCIES

L<DB_File> module is required.

=head1 SEE ALSO

L<Lingua::JA::Romanize::Kana>
L<Lingua::JA::Romanize::Juman>
L<Lingua::JA::Romanize::MeCab>
L<Lingua::ZH::Romanize::Pinyin>
L<Lingua::KO::Romanize::Hangul>

http://www.kawa.net/works/perl/romanize/romanize-e.html

http://openlab.jp/skk/

=head1 AUTHOR

Yusuke Kawasaki http://www.kawa.net/

=head1 COPYRIGHT

Copyright (c) 2006 Yusuke Kawasaki. All rights reserved.

=head1 LICENSE

This dictionary is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either versions 2, or (at your option)
any later version.

=cut
# ----------------------------------------------------------------
package Lingua::JA::Romanize::Japanese;
use strict;
use Carp;
use DB_File;
use Fcntl;
use Lingua::JA::Romanize::Kana;
use vars qw( $VERSION );
$VERSION = "0.14";

my $LINE_MAP = [
    qw(
      a   a   i   i   u   u   e   e   o   o   k   g   k   g   k
      g   k   g   k   g   s   z   s   z   s   z   s   z   s   z   t
      d   t   d   t   t   z   t   d   t   d   n   n   n   n   n   h
      b   p   h   b   p   h   b   p   h   b   p   h   b   p   m   m
      m   m   m   y   y   y   y   y   y   r   r   r   r   r   w   w
      w   w   w   n   b   k   k
      )
];
my $DICT_DB = 'Japanese.bdb';
my $KANAOBJ;

# ----------------------------------------------------------------
sub new {
    my $package = shift;

    my $dbfile = shift || &_detect_sdbm($package);
    Carp::croak "$! - $dbfile\n" unless ( -r $dbfile );

    my $dbhash = {};
    my $flags  = Fcntl::O_RDONLY();
    my $mode   = 0644;
    my $self;
    my $btree = DB_File::BTREEINFO->new();
    tie( %$self, 'DB_File', $dbfile, $flags, $mode, $btree )
      or Carp::croak "$! - $dbfile\n";
    bless $self, $package;

    $KANAOBJ ||= Lingua::JA::Romanize::Kana->new();

    $self;
}

sub char {
    my $self = shift;
    my $char = shift;
    return $self->{$char} if ( exists $self->{$char} && $self->{$char} ne "" );
    $KANAOBJ->char($char);
}

sub chars {
    my $self  = shift;
    my @array = $self->string(shift);
    join( " ", map { $#$_ > 0 ? $_->[1] : $_->[0] } @array );
}

sub string {
    my $self  = shift;
    my $src   = shift;
    my $array = [];

    while ( $src =~ /((?:[\xE0-\xEF][\x80-\xBF]{2})+)|([^\xE0-\xEF]+)/sg ) {
        ### roman
        if ( defined $2 ) {
            push( @$array, [$2] );
            next;
        }
        my $str = $1;
        my $split = [ $str =~ /([\xE0-\xEF][\x80-\xBF]{2})/g ];
        while ( scalar @$split ) {
            ### kana
            if ( $split->[0] =~ /^\xE3/ ) {
                my $kana  = shift @$split;
                my $pair  = [$kana];
                my $roman = $KANAOBJ->char($kana);
                $pair->[1] = $roman if defined $roman;

                # warn "KANA\t[kana=$kana] [roman=$roman]\n";
                push( @$array, $pair );
                next;
            }
            ### kanji
            my $word    = shift @$split;
            my $roman   = $self->char($word);
            my $tryword = $word;
            my $trylist = [];
            while ( scalar @$split ) {
                my $next = $split->[0];

                ### okuri-ari
                if ( $next =~ /^\xE3/ ) {
                    my $okuri = $self->_kana_line($next);
                    if ( exists $self->{ $tryword . $okuri } ) {
                        $roman     = $self->{ $tryword . $okuri };
                        $word      = $tryword;
                        $#$trylist = -1;                             # empty
                          # warn "OKURI\t[kanji=$tryword] [roman=$roman] [okuri=$okuri]\n";
                    }
                }

                last unless exists $self->{ $tryword . $next };
                $tryword .= $next;
                push( @$trylist, shift @$split );
                if ( $self->{$tryword} ne "" ) {
                    $roman     = $self->{$tryword};
                    $word      = $tryword;
                    $#$trylist = -1;                  # empty
                }
            }

            # warn "FIND\t[kanji=$word] [roman=$roman]\n";
            unshift( @$split, @$trylist ) if scalar @$trylist;
            my $pair = defined $roman ? [ $word, $roman ] : [$word];
            push( @$array, $pair );
        }
    }

    $KANAOBJ->normalize($array);
}

sub _kana_line {
    my $self = shift;
    my $char = shift;
    my ( $c1, $c2, $c3 ) = unpack( "C3", $char );
    my $ucs2 =
      ( ( $c1 & 0x0F ) << 12 ) | ( ( $c2 & 0x3F ) << 6 ) | ( $c3 & 0x3F );
    return if ( $ucs2 < 0x3041 );
    return if ( $ucs2 > 0x3093 && $ucs2 < 0x30A1 );
    return if ( $ucs2 > 0x30F6 );
    my $offset = ( $ucs2 - 0x3041 ) % 96;
    $LINE_MAP->[$offset];
}

#   module name to dictionary file path
sub _detect_sdbm {
    my $package = shift;
    my $dbfile = $INC{ join( "/", split( "::", "$package.pm" ) ) };
    $dbfile =~ s#[^/]+$#$DICT_DB#
      or Carp::croak "Invalid module name: $package\n";
    $dbfile;
}

# ----------------------------------------------------------------
1;
